### ISSUE TITLE
NOT IN RANGE exception for smart contract call that returns many values.

### ISSUE THREAD

**Issue by danielefavi** [digitaldonkey/ethereum-php#22]
I am receiving the following exception
**InvalidArgumentException NOT IN RANGE: 100345 ... 082244 > (u)int256**
when I am trying to retrieve information from a smart contract that returns several values.

The exception is raised by the line **92** of the file **SmarContract.php**
`$return = $return->convertByAbi($returnTypeAbi);`

This is the declaration of the function **getProductById** on the smart contract:
```
function getProductById(uint id) public view
returns (
	uint,
	string,
	bytes32,
	uint,
	uint,
	uint[],
	bool
) {
	...
}
```

I think the issue is caused by:
```
$returnTypeAbi = $m->outputs[0]->type;
$return = $return->convertByAbi($returnTypeAbi);
```
where it is trying to convert the response (that is the concatenation of all smartcontract data) into a UINT (that is the first type **$m->outputs[0]** ).
___
FYI:  the ABI is correctly loaded in fact if I dump the `$m->outputs` in the line **91** of the file **SmartContract.php** I can see the value types list declared in the smart contrct function.

**Comment by digitaldonkey** [comment]
uint[] is not supported the library yet. I consider to add https://github.com/sc0Vu/rlp but I need to do some Drupal work first.

**Comment by danielefavi** [comment]
Without `uint[] ` I'm getting the exception
**Call to a member function convertByAbi() on integer**
raised by the function **getMethodParams($m, $values)** of the file **SmartContract.php** on line 138
`$validAbiType = $val->convertByAbi($expectedType);`

This is the new declaration of the function getProductById on the smart contract:

    function getProductById(uint id) public view
    returns (
        uint,
        string,
	bytes32,
        uint,
        uint,
        bool
    ) {
        ...
    }

I would like to know if this exception is related with this issue.
Thank you

**Comment by digitaldonkey** [comment]
"Call to a member function convertByAbi() on integer" means that the integer is not a Instance of EthD. 

Your calls should look similar to
```
$contract->getProductById(new EthQ($yourInt))
```

**Comment by danielefavi** [comment]
The call you mentioned gives me again the exception
NOT IN RANGE: 1277465586268147 ... ... ... 887421292544000 > (u)int256

Instead, it works with smart contract functions that returns one value only. For example this is the function on the smart contract side:
`function getProductName(uint id) public view returns (string) { ... }`

and, in this case, the following call works correctly

    $prdName = $instance->getProductName(new EthQ(123));

I didn't get if I don't use properly your library or the feature with multiple values in return is not supported.

**Comment by tomaszbag** [comment]
Hi, 
What syntax should be used for receiving outbound parameters?

```
function getProductById(uint32 id) public view
returns (
    uint,
    string,
    bytes32,
    uint,
    uint,
    bool
)
```
To call above function I'm using:

`$result = $schemaContract->getProductById(new EthQ($number, ['abi'=>'uint32']));`
	
But 'left' part ($result) is wrong because function getProductById returns many parameters.

Can you help me how above syntax should looks like to get many outbound parameters?

**Comment by digitaldonkey** [comment]
Same problem on eth_protocolVersion()
Seems like I didn't tackle the return values when implementing RLP.  Needs future investigation.

**Comment by tomaszbag** [comment]
hi Thorsten,

Thanks for answer. I had to have a fast solution for returning multiple
parameters from contract function and after many tries with your package I
gave up :)

Next I've installed https://github.com/sc0Vu/web3.php
And there is easier to call all parameters from contract function:

i.e. below:

$contract->at($contractAddress)->call('products', $prdID, function($err,
$result) use ($prdID) {
    if ($err !== null) {
        return $this->fail($err->getMessage());
    }
    if ($result) {
        echo "prdId: ".$result['prdId']."<br />";
        echo "prdName: ".$result['prdName']."<br />";

.....
    }
});

I hope it helps you to fix issue in your package. Thanks again for your
effort.

Best regards,
Tomasz Baglaj



pt., 18.05.2018, 23:04 użytkownik Thorsten Krug <notifications@github.com>
napisał:
